File: Examenes-EDD-6/2024-Enero/Java/TreeBitSet.java
//Student's name: Paco Molina Cuenca
//Student's group: 
//Identity number (DNI if Spanish/passport if Erasmus):

import dataStructures.list.LinkedList;
import dataStructures.list.List;

public class TreeBitSet {
  private static final int BITS_PER_LEAF = LongBits.BITS_PER_LONG;
  
  private interface Tree {
    long size();
    boolean contains(long element, long capacity);
    void add(long element, long capacity);
    List<Long> toList(long capacity);
    Tree cloneTree();
  }

  private final Tree root;
  private final long capacity;

  // returns true if capacity is 64 * 2^n for some n >= 0
  private static boolean isValidCapacity(long capacity) {
    if (capacity <= 0) {
      return false;
    }
    while (capacity > BITS_PER_LEAF) {
      if (capacity % 2 != 0) {
        return false;
      }
      capacity /= 2;
    }
    return capacity == BITS_PER_LEAF;
  }

//-------------------------------------------------------------------
// DO NOT MODIFY ANY CODE ABOVE THIS LINE
//-------------------------------------------------------------------

  private static class Leaf implements Tree {
    private long bitset;
    
    public Leaf(long bitset){
      this.bitset = bitset;
    }

    public long size(){
      return LongBits.countOnes(bitset);
    }

    public boolean contains(long element, long capacity){
      return LongBits.getBit(bitset, (int) element);
    }

    public void add(long element, long capacity){
      bitset = LongBits.setBit(bitset, (int) element);
    }

    public List<Long> toList(long capacity){
      return LongBits.toList(bitset);
    }

    public Tree cloneTree(){
      return new Leaf(bitset);
    }
  }

  private static class Node implements Tree{
    private final Tree left, right;
    
    public Node(Tree left, Tree right){
      this.left = left;
      this.right = right;
    }

    public long size(){
        return left.size() + right.size();
    }

    public boolean contains(long element, long capacity){
      return element < capacity/2 ? left.contains(element, capacity/2) : right.contains(element - capacity/2, capacity/2);
    }

    public void add(long element, long capacity){
      if(element < capacity/2)
        left.add(element, capacity/2);
      else 
        right.add(element - capacity/2, capacity/2);
    }

    public List<Long> toList(long capacity){
      List<Long> res = new LinkedList<>();
      
      for(Long bit : left.toList(capacity/2))
        res.append(bit);

      for(Long bit : right.toList(capacity/2))
        res.append(bit + capacity/2);

      return res;
    }

    public Tree cloneTree(){
      return new Node(left.cloneTree(), right.cloneTree());
    }
  }


  // * Exercise 1 * -
  private static Tree makeTree(long capacity){
    return capacity <= BITS_PER_LEAF ? new Leaf(0) : new Node(makeTree(capacity/2), makeTree(capacity/2));
  }

  // * Exercise 2 * -
  public TreeBitSet(long capacity){
    if(!isValidCapacity(capacity))
      throw new IllegalArgumentException("not valid capacity");

    this.capacity = capacity;
    root = makeTree(capacity);
  }

  // * Exercise 3 * -
  public long capacity(){
    return capacity;
  }

  // * Exercise 4 * -
  private boolean outOfRange(long element){
    return element < 0 || element >= capacity;
  }

  // * Exercise 5 * -
  public long size(){
    return root.size();
  }

  // * Exercise 6 * -
  public boolean isEmpty(){
    return size() == 0;
  }

  // * Exercise 7 * -
  public boolean contains(long element){
    return root.contains(element, capacity);
  }

  // * Exercise 8 * -
  public void add(long element){
    if(outOfRange(element))
      throw new IllegalArgumentException("index out of bounds");

    root.add(element, capacity);
  }

  // * Exercise 9 * -
  public List<Long> toList(){
    return root.toList(capacity);
  }


  //-------------------------------------------------------------------
  // Only for students without continuous assessment
  //-------------------------------------------------------------------
  
  private TreeBitSet(long capacity, Tree root){
    if(!isValidCapacity(capacity))
      throw new IllegalArgumentException("not valid capacity");

    this.capacity = capacity;
    this.root = makeTree(capacity);

    for(Long bit : root.toList(capacity))
      this.add(bit);
  }
  
  // * Exercise 10 * -
  public static TreeBitSet union(TreeBitSet set1, TreeBitSet set2){
	  if(set1.capacity() != set2.capacity())
      throw new IllegalArgumentException("sets have different capacity");

    Tree root = unionRec(set1.root, set2.root);
    TreeBitSet tbs = new TreeBitSet(set1.capacity(), root);

    return tbs;
  }
  
  private static Tree unionRec(Tree tree1, Tree tree2){
	    if(tree1 instanceof Leaf && tree2 instanceof Leaf){
        Leaf leaf1 = (Leaf) tree1;
        Leaf leaf2 = (Leaf) tree2;

        return new Leaf(LongBits.or(leaf1.bitset, leaf2.bitset));

      } else{
        Node node1 = (Node) tree1;
        Node node2 = (Node) tree2;

        return new Node(unionRec(node1.left, node2.left), unionRec(node1.right, node2.right));
      }
  }
  
  // * Exercise 11 * -
  public static TreeBitSet extendedUnion(TreeBitSet set1, TreeBitSet set2){
	    Tree root = extendedUnionRec(set1.root, set1.capacity(), set2.root, set2.capacity());
      long capacity = set1.capacity() <= set2.capacity() ? set2.capacity() : set1.capacity();
      return new TreeBitSet(capacity, root);
  }

  private static Tree extendedUnionRec(Tree tree1, long tam1, Tree tree2, long tam2){
 	    if(tree1 instanceof Leaf && tree2 instanceof Leaf){
        Leaf leaf1 = (Leaf) tree1;
        Leaf leaf2 = (Leaf) tree2;

        return new Leaf(LongBits.or(leaf1.bitset, leaf2.bitset));

      } else if(tree1 instanceof Leaf && tree2 instanceof Node){
        Leaf leaf1 = (Leaf) tree1;
        Node node2 = (Node) tree2;

        return new Node(extendedUnionRec(leaf1, tam1, node2.left, tam2/2), node2.right);

      } else if(tree1 instanceof Node && tree2 instanceof Leaf){
        Node node1 = (Node) tree1;
        Leaf leaf2 = (Leaf) tree2;
        
        return new Node(extendedUnionRec(node1.left, tam1/2, leaf2, tam2), node1.right);

      } else{
        Node node1 = (Node) tree1;
        Node node2 = (Node) tree2;

        return new Node(extendedUnionRec(node1.left, tam1/2, node2.left, tam2/2), extendedUnionRec(node1.right, tam1/2, node2.right, tam2/2));
      }
  }

  //-------------------------------------------------------------------
  // Basic program for testing your implementation
  //-------------------------------------------------------------------
  public static void main(String[] args) {
    TreeBitSet set = new TreeBitSet(64*2*2*2);
    set.add(0);
    set.add(1);
    set.add(2);
    set.add(3);
    set.add(4);
    set.add(5);
    set.add(6);
    set.add(7);
    set.add(8);
    set.add(9);
    set.add(10);
    set.add(120);
    set.add(128);
    set.add(270);

    System.out.println(set.size());

    System.out.println(set.contains(6));
    System.out.println(set.contains(128));
    System.out.println(set.contains(270));

    System.out.println(set.contains(11));
    System.out.println(set.contains(272));

    System.out.println(set.toList());
  }
}

File: Examenes-EDD-7/2024-Enero/Haskell/DataStructures/Set/TreeBitSet.hs
--Student's name: Paco Molina Cuenca
--Student's group: 
--Identity number (DNI if Spanish/passport if Erasmus):

module DataStructures.Set.TreeBitSet (
  TreeBitSet,
  empty,
  size,
  capacity,
  isEmpty,
  contains,
  add,
  toList

  ) where

import Test.QuickCheck
import qualified DataStructures.Set.IntBits as IntBits
import Data.List((\\))


data Tree = Leaf Int | Node Tree Tree deriving Show

data TreeBitSet = TBS Int Tree deriving Show

bitsPerLeaf :: Int
bitsPerLeaf = 64

-- returns true if capacity is 64 * 2^n for some n >= 0
isValidCapacity :: Int -> Bool
isValidCapacity capacity
  | capacity <= 0           = False
  | capacity == bitsPerLeaf = True
  | m == 0                  = isValidCapacity half
  | otherwise               = False
  where
    (half, m) = divMod capacity 2

-- =================================================================

-- | Función para crear un árbol con la capacidad dada
makeTree :: Int -> Tree
makeTree tam
  | tam <= bitsPerLeaf = Leaf 0 
  | otherwise          = Node (makeTree half) (makeTree half)
    where 
      half = tam `div` 2

-- | Función para crear un TreeBitSet vacío con la capacidad dada
empty :: Int -> TreeBitSet
empty tam
  | tam < 0                   = error "capacity must be positive"
  | not (isValidCapacity tam) = error "capacity must be 64 multiplied by a power of 2"
  | otherwise                 = TBS tam (makeTree tam)

-- | Función para obtener el tamaño de un TreeBitSet
size :: TreeBitSet -> Int
size (TBS _ tree) = sizeRec tree 
  where 
    sizeRec (Leaf bitSet) = IntBits.countBits bitSet 
    sizeRec (Node lt rt)  = sizeRec lt + sizeRec rt

-- | Función para obtener la capacidad de un TreeBitSet
capacity :: TreeBitSet -> Int
capacity (TBS tam _) = tam

-- | Función para verificar si un TreeBitSet está vacío
isEmpty :: TreeBitSet -> Bool
isEmpty tbs = size tbs == 0

-- | Función para verificar si un elemento está fuera del rango de un TreeBitSet
outOfRange :: TreeBitSet -> Int -> Bool
outOfRange (TBS tam _) i = i < 0 || i >= tam

-- | Función para verificar si un elemento está contenido en un TreeBitSet
contains :: Int -> TreeBitSet -> Bool
contains i tbs@(TBS tam tree) 
  | outOfRange tbs i = False 
  | otherwise        = containsRec i tam tree 
    where 
      containsRec i _ (Leaf bitSet)  = IntBits.bitValue i bitSet
      containsRec i tam (Node lt rt)
        | i < half  = containsRec i half lt 
        | otherwise = containsRec (i - half) half rt
            where 
              half = tam `div` 2

-- | Función para agregar un elemento a un TreeBitSet
add :: Int -> TreeBitSet -> TreeBitSet
add i tbs@(TBS tam tree) 
  | outOfRange tbs i = error "add on tbs with index out of bounds"
  | otherwise        = TBS tam (addRec i tam tree)
    where 
      addRec i _ (Leaf bitSet)  = Leaf (IntBits.setBit' i bitSet)
      addRec i tam (Node lt rt)
        | i < half  = Node (addRec i half lt) rt
        | otherwise = Node lt (addRec (i - half) half rt)
          where 
            half = tam `div` 2 

-- | Función para convertir un TreeBitSet a una lista de elementos
toList :: TreeBitSet -> [Int]
toList (TBS tam tree) = toListRec 0 tam tree 
  where 
    toListRec offSet _ (Leaf bitSet)  = map (+offSet) (IntBits.toList bitSet)
    toListRec offSet tam (Node lt rt) = toListRec offSet half lt ++ toListRec (offSet + half) half rt 
      where 
        half = tam `div` 2

-- | Función para realizar la unión de dos TreeBitSets
union :: TreeBitSet -> TreeBitSet -> TreeBitSet
union (TBS tam1 tree1) (TBS tam2 tree2)
  | tam1 /= tam2 = error "union on TBS with different shape"
  | otherwise    = TBS tam1 (unionRec tree1 tree2)
    where 
      unionRec (Leaf bitSet1) (Leaf bitSet2) = Leaf (foldr IntBits.setBit' bitSet2 (IntBits.toList bitSet1))  
      unionRec (Node lt1 rt1) (Node lt2 rt2) = Node (unionRec lt1 lt2) (unionRec rt1 rt2)

-- | Función para realizar la unión extendida de dos TreeBitSets
extendedUnion :: TreeBitSet -> TreeBitSet -> TreeBitSet
extendedUnion tbs1@(TBS tam1 tree1) tbs2@(TBS tam2 tree2)
  | tam1 < tam2  = TBS tam2 (extendedUnionRec tam1 tree1 tam2 tree2) 
  | tam1 == tam2 = union tbs1 tbs2
  | otherwise    = TBS tam1 (extendedUnionRec tam1 tree1 tam2 tree2)
    where 
      extendedUnionRec tam1 (Leaf bitSet1) tam2 (Leaf bitSet2) = Leaf (foldr IntBits.setBit' bitSet2 (IntBits.toList bitSet1)) 
       where 
          half1 = tam1 `div` 2 
          half2 = tam2 `div` 2
      extendedUnionRec tam1 (Leaf bitSet1) tam2 (Node lt2 rt2) = Node (extendedUnionRec tam1 (Leaf bitSet1) half2 lt2) rt2
        where 
          half1 = tam1 `div` 2 
          half2 = tam2 `div` 2
      extendedUnionRec tam1 (Node lt1 rt1) tam2 (Leaf bitSet2) = Node (extendedUnionRec half1 lt1 tam2 (Leaf bitSet2)) rt1
        where 
          half1 = tam1 `div` 2 
          half2 = tam2 `div` 2
      extendedUnionRec tam1 (Node lt1 rt1) tam2 (Node lt2 rt2) = Node (extendedUnionRec half1 lt1 half2 lt2) (extendedUnionRec half1 rt1 half2 rt2)
        where 
          half1 = tam1 `div` 2 
          half2 = tam2 `div` 2


-- =========================================

instance Arbitrary TreeBitSet where
    arbitrary = do
        exponent <- chooseInt (0, 2)
        let capacity = bitsPerLeaf * 2^exponent
        elements <- listOf (chooseInt (0, capacity-1))
        return (foldr add (empty capacity) elements)
            where
                chooseInt :: (Int, Int ) -> Gen Int 
                chooseInt (x,y) = choose (x,y)        

instance Eq TreeBitSet where
  tbs1 == tbs2 = toInts tbs1 == toInts tbs2
    where
      toInts (TBS _ t) = reverse . dropWhile (==0) . reverse $ toInts' t
      toInts' (Leaf b) = [b]
      toInts' (Node lt rt) = toInts' lt ++ toInts' rt

-- =========================================


validExponent e = e >= 0 && e <= 5

validElement e (TBS c t) = e >= 0 && e < c

-- Axioms for basic set of operations
ax1 e = validExponent e ==> isEmpty emptySet
  where emptySet = empty (64*2^e)

ax2 x s = validElement x s ==> not (isEmpty (add x s))

ax3 e x = validExponent e ==> not (contains x emptySet)
  where emptySet = empty (64*2^e)

ax4 x y s = validElement y s ==> contains x (add y s) == (x==y) || contains x s

ax5 e = validExponent e ==> size emptySet == 0
  where emptySet = empty (64*2^e)

ax6 x s = contains x s  ==> size (add x s) == size s
ax7 x s = validElement x s && not (contains x s) ==> size (add x s) == 1 + size s

ax11 e = validExponent e ==> null (toList emptySet)
  where emptySet = empty (64*2^e)

ax12 x s = contains x s  ==> toList (add x s) == toList s

ax13 x s = validElement x s && not (contains x s) ==> x `elem` xs && (xs \\ [x]) `sameElements` toList s
  where
    xs = toList (add x s)
    sameElements xs ys = null (xs \\ ys) && null (ys \\ xs)


type Elem = Int

setAxioms = do
  quickCheck (ax1 :: Int -> Property)
  quickCheck (ax2 :: Elem -> TreeBitSet -> Property)
  quickCheck (ax3 :: Int -> Elem -> Property)
  quickCheck (ax4 :: Elem -> Elem -> TreeBitSet -> Property)
  quickCheck (ax5 :: Int -> Property)
  quickCheck (ax6 :: Elem -> TreeBitSet -> Property)
  quickCheck (ax7 :: Elem -> TreeBitSet -> Property)
  quickCheck (ax11 :: Int -> Property)
  quickCheck (ax12 :: Elem -> TreeBitSet -> Property)
  quickCheck (ax13 :: Elem -> TreeBitSet -> Property)
